src/datahike/writer.cljc

```diff
@@ -1,75 +1,143 @@
             [datahike.core]
             [datahike.writing :as w]
-            [datahike.tools :as dt :refer [throwable-promise]]
-            [clojure.core.async :refer [chan close! promise-chan put! go go-loop <!]])
+            [datahike.tools :as dt :refer [throwable-promise get-time-ms]]
+            [clojure.core.async :refer [chan close! promise-chan put! go go-loop <! >! poll! buffer timeout]])
   (:import [clojure.lang ExceptionInfo]))
 
 (defprotocol PWriter
   (-dispatch! [_ arg-map] "Returns a channel that resolves when the transaction finalizes.")
   (-shutdown [_] "Returns a channel that resolves when the writer has shut down.")
   (-streaming? [_] "Returns whether the transactor is streaming updates directly into the connection, so it does not need to fetch from store on read."))
 
-(defrecord LocalWriter [queue thread streaming?]
+(defrecord LocalWriter [thread streaming? transaction-queue-size commit-queue-size
+                        transaction-queue commit-queue]
   PWriter
   (-dispatch! [_ arg-map]
     (let [p (promise-chan)]
-      (put! queue (assoc arg-map :callback p))
+      (put! transaction-queue (assoc arg-map :callback p))
       p))
   (-shutdown [_]
-    (close! queue)
+    (close! transaction-queue)
     thread)
   (-streaming? [_] streaming?))
 
+(def ^:const DEFAULT_QUEUE_SIZE 120000)
+
+;; minimum wait time between commits in ms
+;; this reduces write pressure on the storage
+(def ^:const DEFAULT_COMMIT_WAIT_TIME 0) ;; in ms
+
 (defn create-thread
   "Creates new transaction thread"
-  [connection queue write-fn-map]
-  (thread-try
-   S
-   (go-loop []
-     (if-let [{:keys [op args callback] :as invocation} (<! queue)]
-       (do
-         (let [op-fn (write-fn-map op)
-               res   (try
-                       (apply op-fn connection args)
-                       ;; Only catch ExceptionInfo here (intentionally rejected transactions).
-                       ;; Any other exceptions should crash the writer and signal the supervisor.
-                       (catch Exception e
-                         (log/errorf "Error during invocation" invocation e)
-                         ;; take a guess that a NPE was triggered by an invalid connection
-                         (if (= (type e) NullPointerException)
-                           (ex-info "Null pointer encountered in invocation. Connection may have been invalidated, e.g. through db deletion, and needs to be released everywhere."
-                                    {:type       :writer-error-during-invocation
-                                     :invocation invocation
-                                     :connection connection
-                                     :error      e})
-                           e)))]
-           (when (some? callback)
-             (put! callback res)))
-         (recur))
-       (log/debug "Writer thread gracefully closed")))))
+  [connection write-fn-map transaction-queue-size commit-queue-size commit-wait-time]
```

Comment:
This is the main added logic. It first transacts in a loop and then commits in the second loop. Transactions are grouped in commits and they all see the same db-after containing their transaction and their individual datoms added and retracted.